.github/workflows/cancel-dupliate-workflow-runs.yml

```diff
@@ -0,0 +1,39 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+name: "Cancel Duplicates"
+on:
+  workflow_run:
+    workflows: 
+      - 'Java CI'
+      - 'Python CI'
+    types: ['requested']
+
+jobs:
+  cancel-duplicate-workflow-runs:
+    name: "Cancel duplicate workflow runs"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: potiuk/cancel-workflow-runs@4723494a065d162f8e9efd071b98e0126e00f866 # @master
+        name: "Cancel duplicate workflow runs"
+        with:
+          cancelMode: allDuplicates
+          token: ${{ secrets.GITHUB_TOKEN }}
```

Comment:
Why do we need to explicitly pass the token?

Reply:
I believe the token is needed so that this workflow can cancel other workflows.

It seems to be required by a lot of workflows,[ including the labeler workflow](https://github.com/apache/iceberg/blob/master/.github/workflows/labeler.yml#L27-L29). According to the docs, it's also technically accessible on the `github` context even if not explicitly passed.

Of note, according to the docs linked below, Github **creates a token when the action is started and that [token is only valid for the lifetime of the actions run.]**(https://docs.github.com/en/actions/reference/authentication-in-a-workflow#about-the-github_token-secret)

```
Before each job begins, GitHub fetches an installation access token for the job.
The token expires when the job is finished.
```

Here's a matrix of the [default permissions based on what's taking place](https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token) (for us, this would fall under `actions`).

![image](https://user-images.githubusercontent.com/9833362/130371781-05d54dbb-dada-419b-8a5d-deb2f5c1ef43.png)

Reply:
Also of note, permissions can be configured for the token that's grabbed for each action.

Here's an example using the labeler action (that we also use) from those above links:

```yaml
name: Pull request labeler

on: [ pull_request_target ]

permissions:
  contents: read
  pull-requests: write
```

Appears that the generated token will only have read/write or read access to the `scope` defined in `permissions`.

Here's also[ an article on not preventing pwn requests from Github Actions](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/) (which I still need to read in further depth).

Reply:
Finally, and _**perhaps most importantl**_ , I just noticed that Github has a new-ish feature, that is documented as  _beta_, called `concurrency` that would allow us to do this natively within the GithubAction (no additional workflow - same permissions as the running workflow itself presumably).

Announcement summary: https://github.blog/changelog/2021-04-19-github-actions-limit-workflow-run-or-job-concurrency/

Docs: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#concurrency (which is where it's mentioned as technically still beta).

Do we potentially want to consider the native route? It was announced or released in April, so the action being used here is currently marked as planned for deprecation / deprecated (but with the caveat that `concurrency` is still beta). This particular Github Action is still being used in Spark.

I'm willing to volunteer to help triage issues with `concurrency` with as much immediacy as I can afford if something "beta" about it changes that breaks our workflows (especially if we allow for simply removing the `concurrency` section to give time to fix it without blocking all PR runs).

Perhaps we try `concurrency` on one github action and use this workflow for the rest as it's stable (and used by Spark currently)? 100% open to suggestions here.

I'm also going to add a `permissions` section to the labeler in another PR / issue to try that out. 🙂

Reply:
I've opened a PR for adding a `permissions` section to the labeler (which I've tested in my fork): https://github.com/apache/iceberg/pull/3007

Reply:
Thanks for looking into this. Let's go with this for now since it is working and we can update to the native one later.

Reply:
Cool sounds good. I'll keep my eye on `concurrency` in general.